### ISSUE TITLE
Info on how to test handlers?

### ISSUE THREAD

**Issue by david4shure** [guregu/kami#29]
So I have a project written using the kami web framework and I am beginning to write tests for each of my handlers in turn, is there any documentation on best practices for testing and how to use kami handlers during those tests? I have looked specifically at this guide https://elithrar.github.io/article/testing-http-handlers-go/ and was wondering if you had any guides on how specifically to accomplish these styles of functional handler tests with kami.

**Comment by guregu** [comment]
Hello,
The more important advice I can give is this: make sure you run your tests through kami instead of calling your handler directly, if you want your middleware to run and `kami.Param()` to work.

In the blog post you linked they have code like:
```go
// Our handlers satisfy http.Handler, so we can call their ServeHTTP method 
// directly and pass in our Request and ResponseRecorder.
handler.ServeHTTP(rr, req)
```

Instead, do something like this. Creating a new mux for every test would be the most 'pure' way to test your project.
```go
// set up our mux, you can move this to its own function somewhere
mux := kami.New()
mux.Context = myContext
mux.Get("/mypath", handler)
mux.Use("/mypath", middleware)
// set up the httptest.ResponseRecorder etc
mux.ServeHTTP(rr, req) // ← this is the important part, make sure you run the request through kami instead of calling the handler directly
```

Personally, I prefer using the global mux. I set up my routes and middleware in the package's `init()` function, like so:

```go
func init() {
	kami.Use("/", Authenticate)

	kami.Use("/user/", RequireLogin)
	kami.Get("/user/", UserProfile)
	kami.Post("/user/", EditProfile)
}
```

This way, during testing your routes will always be the same as production. Then, in tests I call the global mux directly like `kami.Handler().ServeHTTP(rr, req)` to serve each request. If you don't like using `init()`, you could return a mux instead like `func newMux() *kami.Mux` and use that for your tests.

To handle context for testing, you could create a new mux each time and change the mux's context. If all your tests can share the same test context, you could simply set `kami.Context` to your test context. 

For some of my projects that have a common context for tests, I have a package called `tests` or `testhelper` with code like:

```go
package tests

import (
	"github.com/guregu/kami"
	"golang.org/x/net/context"

	"github.com/my-company/my-project/log"
)

func init() {
	// build base context for incoming requests
	ctx := context.Background()
	ctx = log.ConnectForTesting(ctx)
	kami.Context = ctx
}

func Context() context.Context {
	return kami.Context
}
```

This way if you import this for your tests it automatically sets everything up. I realize not everyone like init functions, so YMMV.

**tl;dr**: You can follow the same steps as that blog post, but make sure you call your handlers through kami (either via `kami.Handler().ServeHTTP()` or `mux.ServeHTTP()`) to make path parameters work and middleware run.

I hope this helps, let me know if you have any other questions.

**Comment by david4shure** [comment]
This is very valuable information. I appreciate you taking the time to put it together, would it be okay to add this to your readme?

**Comment by guregu** [comment]
The readme file is getting pretty bloated so maybe it'd be a good idea to start moving stuff to the wiki. I think anyone can edit the wiki right now, so if you make a testing wiki page I'll add a link to it in the readme.

**Comment by david4shure** [comment]
Sounds good, I'll add that to the wiki when I get a chance 👍
